### PR TITLE
fix: utility for marking suggestions as covered

### PR DESCRIPTION
**Pull request by dipratap**
Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues

Thanks for contributing!

### CODE REVIEW

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @dipratap,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - tests assert `saveMany` behavior that `saveSuggestions` does not produce given the current `PARALLEL_SAVE_THRESHOLD`, and the new public method lacks error handling present in the original code path.
**Complexity: MEDIUM** - medium-size refactoring across 3 files.
_Changes: Extracts suggestion-coverage filtering into a shared `findCoveredSuggestions` utility and adds a new `markPatternCoveredSuggestions` public method on TokowakaClient (3 files)._

## Must fix before merge

1. [Critical] Tests assert `saveMany` is called but `saveSuggestions` takes the individual-save/SQS path with the current threshold - test mocks also lack `.save()` method, causing TypeError - `packages/spacecat-shared-tokowaka-client/test/index.test.js:181`

   Tests assert `client.dataAccess.Suggestion.saveMany.calledOnceWith(covered)` but `saveSuggestions` with the current `PARALLEL_SAVE_THRESHOLD` value enters the individual-save path (`Promise.allSettled(suggestions.map(s => s.save()))`) rather than calling `saveMany`. The `mkCovered` mock objects lack a `.save()` method, so this path throws `TypeError: s.save is not a function`.

   Fix options:
   - Add `save: sinon.stub().resolves()` to `mkCovered` and assert on individual `.save()` calls instead of `saveMany`.
   - Set up `client.sqs` and `client.importWorkerQueueUrl` in the test fixture (as the bulk-update test already does) so the SQS path is taken, and assert on `sqs.sendMessage`.
   - If `saveMany` is the intended path for small batches, verify the threshold value allows it.

2. [Important] Missing error handling around `saveSuggestions` - the original `deployPatternSuggestion` wraps saves in try/catch with a warning log; the new method propagates unhandled - `packages/spacecat-shared-tokowaka-client/src/index.js:1766`

   The original code wraps `saveSuggestions` in a try/catch that logs a warning and continues gracefully on failure. The new `markPatternCoveredSuggestions` has no equivalent guard - a save failure (SQS down, DB error) propagates as an unhandled rejection, which could abort the caller's entire batch operation.

   Fix: Either add a try/catch with a warning log (matching the established resilience pattern), or document in the JSDoc that callers are responsible for error handling (`@throws` on save failure).

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: Commit type `fix:` may be more appropriate as `feat:` since a new public method and exported utility are added (conventional commits required per CLAUDE.md) - PR title
- suggestion: Add dedicated unit tests for `findCoveredSuggestions` in `test/utils/suggestion-utils.test.js` to mirror `src/` structure per CLAUDE.md convention and cover branches not exercised through the client (excludeIds, edgeDeployed skip, invalid regex patterns) - `packages/spacecat-shared-tokowaka-client/test/utils/suggestion-utils.test.js`
- nit: `patternSuggestion.getData()` called twice in `markPatternCoveredSuggestions` - store once and reference for both `allowedRegexPatterns` and `isDomainWide` - `packages/spacecat-shared-tokowaka-client/src/index.js:1741`
- suggestion: The `[edge-deploy]` log prefix in `markPatternCoveredSuggestions` may be misleading when called from the launch-time cover-marking flow; consider a more generic prefix or a caller-supplied label - `packages/spacecat-shared-tokowaka-client/src/index.js:1783`

</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 3m 12s | <b>Cost:</b> $7.62 | <b>Commit:</b> <code>3c404b5902e6f9a1cbde2ac9eb8a729583c0d8ac</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: packages/spacecat-shared-tokowaka-client/test/index.test.js [hunk 1]
```diff
@@ -147,6 +147,98 @@ describe('TokowakaClient', () => {
...
+
+    it('warns and returns [] instead of throwing when saveSuggestions fails', async () => {
+      client.dataAccess.Suggestion.saveMany.rejects(new Error('DB down'));
+      const covered = [mkCovered('a')];
+
+      const result = await client.markPatternCoveredSuggestions(mkPattern(true), covered, 'site-1');
+
+      expect(result).to.deep.equal([]);
+      expect(client.log.warn.calledWithMatch(/Failed to mark covered suggestions/)).to.be.true;
+    });
```
Comment (MysticatBot) [inline packages/spacecat-shared-tokowaka-client/test/index.test.js:239]: suggestion: `calledOnceWith(covered)` validates the first arg but does not verify `{ chunkSize: 25 }` as the second. This couples the test to `saveSuggestions` internals without fully asserting the contract. Consider `calledOnceWithExactly(covered, sinon.match({ chunkSize: 25 }))` if you want to lock the chunking behavior, or drop the `saveMany` assertion entirely and rely on the observable outcome (returned array, mutated data) if you want the test decoupled from the save mechanism.

File: packages/spacecat-shared-tokowaka-client/test/index.test.js [hunk 2]
```diff
@@ -147,6 +147,98 @@ describe('TokowakaClient', () => {
...
+      const [queueUrl, payload] = client.sqs.sendMessage.firstCall.args;
+      expect(queueUrl).to.equal('https://sqs.test/import-worker-queue');
+      expect(payload).to.include({ type: 'suggestion-bulk-update', siteId: 'site-1' });
+      expect(payload.set).to.deep.equal({ coveredByDomainWide: 'dw-1' });
+      expect(payload.updatedBy).to.equal('tester');
+      expect(client.dataAccess.Suggestion.saveMany.called).to.be.false;
+    });
+
+    it('warns and returns [] instead of throwing when saveSuggestions fails', async () => {
+      client.dataAccess.Suggestion.saveMany.rejects(new Error('DB down'));
```
Comment (MysticatBot) [inline packages/spacecat-shared-tokowaka-client/test/index.test.js:232]: suggestion: This test validates `type`, `siteId`, `set`, and `updatedBy` but does not assert `suggestionIds` in the SQS payload. That field tells the import worker which suggestions to update - verifying `expect(payload.suggestionIds).to.have.length(1701)` would close the gap.

**Review by MysticatBot — APPROVED** [review]
Hey @dipratap,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - prior blocking findings addressed, extraction is clean and well-tested.
**Complexity: MEDIUM** - medium-size refactoring with new public API method.
_Changes: Extracts suggestion-coverage filtering into a shared `findCoveredSuggestions` utility and adds a new `markPatternCoveredSuggestions` public method on TokowakaClient (3 files)._

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: Test asserts `saveMany` but does not verify the `chunkSize` option passed as second arg - coupling to `saveSuggestions` internals makes the assertion fragile if routing logic changes - `packages/spacecat-shared-tokowaka-client/test/index.test.js:239`
- suggestion: Add dedicated unit tests for `findCoveredSuggestions` in `test/utils/suggestion-utils.test.js` to mirror `src/` structure per CLAUDE.md convention and cover edge cases (excludeIds, edgeDeployed skip, pattern-covers-pattern logic) independently - `packages/spacecat-shared-tokowaka-client/src/utils/suggestion-utils.js`
- suggestion: SQS bulk-update test should assert `suggestionIds` is present with expected length (1701) - this is the most important field for the import worker to know what to update - `packages/spacecat-shared-tokowaka-client/test/index.test.js:280`
- nit: JSDoc for `markPatternCoveredSuggestions` does not describe which data field gets mutated (`coveredByDomainWide` vs `coveredByPattern`) or how it is determined - a one-line note would complete the contract for consumers - `packages/spacecat-shared-tokowaka-client/src/index.js:1732`

</details>

## Previously flagged, now resolved
- Test assertion mismatch with `saveMany` - confirmed correct: 2-item batch routes through `saveMany` (threshold is 1700); mocks now include `.save()` stub defensively.
- Missing error handling around `saveSuggestions` - try/catch with `log.warn` and `return []` now wraps the save call, matching the established resilience pattern.

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 5s | <b>Cost:</b> $5.36 | <b>Commit:</b> <code>46cc13777534c4ef2cc2b6be182c8f5dbcc33e96</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
